File: backend/services/firebaseService.js
import path from "path";
import { v4 as uuidv4 } from "uuid";

import { initFirebase } from "../config/firebase.js";

export async function uploadPdfForUser({ userId, pdfPath }) {
  try {
    const admin = await initFirebase();
    const bucket = admin.storage().bucket();

    const objectName = `users/${userId}/${uuidv4()}.pdf`;

    await bucket.upload(pdfPath, {
      destination: objectName,
      metadata: {
        contentType: "application/pdf",
        cacheControl: "public, max-age=31536000"
      }
    });

    const file = bucket.file(objectName);

    // Prefer a long-lived signed URL to avoid requiring bucket-wide public ACLs.
    const [signedUrl] = await file.getSignedUrl({
      action: "read",
      expires: "2100-01-01T00:00:00Z"
    });

    return signedUrl;
  } catch (err) {
    throw Object.assign(new Error(`Firebase upload error: ${err.message}`), { statusCode: err.statusCode || 502 });
  }
}

export async function uploadRecordingForUser({ userId, recordingPath }) {
  try {
    const admin = await initFirebase();
    const bucket = admin.storage().bucket();

    const objectName = `users/${userId}/${uuidv4()}.mp4`;

    await bucket.upload(recordingPath, {
      destination: objectName,
      metadata: {
        contentType: "video/mp4",
        cacheControl: "public, max-age=31536000"
      }
    });

    const file = bucket.file(objectName);

    const [signedUrl] = await file.getSignedUrl({
      action: "read",
      expires: "2100-01-01T00:00:00Z"
    });

    return signedUrl;
  } catch (err) {
    throw Object.assign(new Error(`Firebase upload error: ${err.message}`), { statusCode: err.statusCode || 502 });
  }
}


export async function getUserData(userId) {
  try {
    const admin = await initFirebase();
    const db = admin.firestore();
    const userDoc = await db.collection("users").doc(userId).get();
    if (!userDoc.exists) return null;
    return userDoc.data();
  } catch (err) {
    console.error(`[FIREBASE] Error fetching user ${userId}:`, err);
    return null;
  }
}

export async function getUserRecordingCount(userId) {
  try {
    const admin = await initFirebase();
    const db = admin.firestore();
    const snap = await db.collection("recordings")
      .where("userId", "==", userId)
      .get();
    return snap.size;
  } catch (err) {
    console.error(`[FIREBASE] Error counting recordings for ${userId}:`, err);
    return 0;
  }
}
export async function updateJobStatus(jobId, data) {
  try {
    const admin = await initFirebase();
    const db = admin.firestore();
    await db.collection("processing_jobs").doc(jobId).set({
      ...data,
      updatedAt: admin.firestore.FieldValue.serverTimestamp()
    }, { merge: true });
    console.log(`[FIREBASE] Job ${jobId} updated: ${data.status}`);
  } catch (err) {
    console.error(`[FIREBASE] Error updating job ${jobId}:`, err);
  }
}
export async function downloadFileFromStorage(storagePath, localPath) {
  try {
    const admin = await initFirebase();
    const bucket = admin.storage().bucket();
    
    // If it's a full URL, we might need to extract the path, 
    // but better to have the client send the storage path directly.
    await bucket.file(storagePath).download({ destination: localPath });
    console.log(`[FIREBASE] Downloaded ${storagePath} to ${localPath}`);
  } catch (err) {
    console.error(`[FIREBASE] Download failed for ${storagePath}:`, err);
    throw new Error(`Firebase download error: ${err.message}`);
  }
}

File: backend_python/services/firebase_service.py
import os
from datetime import datetime, timedelta
from config.firebase_config import get_db, get_bucket, init_firebase
from firebase_admin import firestore

def update_job_status(job_id: str, data: dict):
    """
    Updates the status of a processing job in Firestore.
    """
    try:
        db = get_db()
        data['updatedAt'] = firestore.SERVER_TIMESTAMP
        db.collection("processing_jobs").document(job_id).set(data, merge=True)
        print(f"[FIREBASE] Job {job_id} updated: {data.get('status')}")
    except Exception as e:
        print(f"[FIREBASE ERROR] Failed to update job {job_id}: {str(e)}")

import requests

def download_file_from_storage(ref: str, local_path: str):
    """
    Downloads a file from either a full URL (HTTPS) or a Firebase Storage path.
    """
    try:
        # Case 1: Full URL (starts with http)
        if ref.startswith("http"):
            print(f"[FIREBASE] Downloading from URL: {ref[:50]}...", flush=True)
            response = requests.get(ref, stream=True, timeout=60)
            response.raise_for_status()
            with open(local_path, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
            print(f"[FIREBASE] Download complete via URL: {local_path}")
            
        # Case 2: Relative Storage Path (e.g. recordings/user/...)
        else:
            print(f"[FIREBASE] Downloading from Storage Path: {ref}", flush=True)
            bucket = get_bucket()
            blob = bucket.blob(ref)
            blob.download_to_filename(local_path)
            print(f"[FIREBASE] Download complete via Storage SDK: {local_path}")
            
    except Exception as e:
        print(f"[FIREBASE ERROR] Download failed for '{ref[:50]}': {str(e)}")
        raise e

def upload_file_to_storage(local_path: str, destination_path: str, content_type: str) -> str:
    """
    Uploads a local file to Firebase Storage and returns a signed URL.
    """
    try:
        bucket = get_bucket()
        blob = bucket.blob(destination_path)
        blob.upload_from_filename(local_path, content_type=content_type)
        
        # Generate a long-lived signed URL
        url = blob.generate_signed_url(
            expiration=timedelta(days=3650), # ~10 years
            method='GET'
        )
        return url
    except Exception as e:
        print(f"[FIREBASE ERROR] Upload failed for {local_path}: {str(e)}")
        raise e

def get_user_data(user_id: str):
    try:
        db = get_db()
        doc = db.collection("users").document(user_id).get()
        return doc.to_dict() if doc.exists else None
    except Exception as e:
        print(f"[FIREBASE ERROR] Failed to get user data for {user_id}: {str(e)}")
        return None
